reference/src/glossary.md

```diff
@@ -1,5 +1,48 @@
 ## Glossary
 
+#### Aliasing
+
+(Please note: a full aliasing model for Rust has not yet been constructed, but
+at the moment we can give the following definition.)
+
+*Aliasing* is any time one pointer or reference points to a "span" of memory
+that overlaps with the span of another pointer or reference. A span of memory is
+similar to how a slice works: there's a base byte address as well as a length in
+bytes.
+
+Consider the following example:
+
+```rust
+fn main() {
+    let u: u64 = 7_u64;
+    let r: &u64 = &u;
+    let s: &[u8] = unsafe {
+        core::slice::from_raw_parts(&u as *const u64 as *const u8, 8)
+    };
+    let (head, tail) = s.split_first().unwrap();
+}
+```
+
+In this case, both `r` and `s` alias each other, since they both point to all of
+the bytes of `u`.
+
+However, `head` and `tail` do not alias each other: `head` points to the first
+byte of `u` and `tail` points to the other seven bytes of `u` after it.
```

Comment:
I think it is worth stating the relationship between `head`/`tail` and `s` here, e.g., do `head` and `tail` alias `s` ?

For example, given:

```rust
    let mut x = 32_i32;
    let y = &mut x;
    let z = y as *mut _;
    unsafe {
        *z = 3;
    }
    *y = 42;
```

When does `z` alias `y` and whether that is UB ? I suppose that without an aliasing model, it might not be possible to tell what's exactly UB, but it would be worth it to write down what precisely is unspecified because of the lack of an aliasing model and showing a small example.

Reply:
This particular example seems similar to a thing @RalfJung has explained before... I _think_ that `z` counts as a further borrow on `y`, so... something.

I mean it's clear that `y` and `z` point to the same memory span. It's not clear what that means in terms of aliasing/ownership rules.

Reply:
> I think it is worth stating the relationship between head/tail and s here, e.g., do head and tail alias s ?

They do, and that is worth stating.

However, the point of this definition is to define when aliasing *happens*, not when it is *allowed*.

Reply:
added

Reply:
Thanks!

@gnzlbg do you agree with this? I don't think we should even try to define "allowed vs disallowed aliasing" in the glossary at this point.

At best we could link to `wip/stacked-borrows.md`.

Reply:
> I don't think we should even try to define "allowed vs disallowed aliasing" in the glossary at this point.

Ok, sounds good. I guess what confused me is that at the beginning an "aliasing model" is mentioned, and it wasn't fully clear to me how far this aliasing definition should go.

I think linking to the stacked-borrows wip document as an example of one potential aliasing model would help.

Reply:
alright, can do after work

Reply:
If you mean https://github.com/Lokathor/unsafe-code-guidelines/blob/lokathor/wip/stacked-borrows.md the start of that says that it's not endorsed by the UCG WG. We could make it endorsed, and link to it, but that should be a separate PR if so.

Reply:
The idea was to link to it as an *example* of what an aliasing model *could* look like, not as *the* aliasing model.

Reply:
Alright, check now